### PR TITLE
renamed module tls client auth params to avoid overlaping with fetch_url

### LINKED ISSUE #59513: pulp_repo - using repo with tls client auth fails
##### SUMMARY
When you wan't to create / update a repo which needs tls client auth for sync (like redhat repos)
this breaks because this module uses the following vars:
```python
ca_cert
client_cert
client_key
```
which overlap with the fetch_url vars which also use those.
So the api calls to your local pulp server are also made with tls client auth and
not only for the remote repo.

This only affects users who wan't to sync RedHat repos which needs TLS client auth.

The bug happens only in call **server.set_repo_list()** of the module:
```python
    server = pulp_server(module, pulp_host, repo_type, wait_for_completion=wait_for_completion)
    server.set_repo_list()
    repo_exists = server.check_repo_exists(repo)

```

the call to **fetch_url** has overlapping vars the module defines:
```python
   def set_repo_list(self):
        url = "%s/pulp/api/v2/repositories/?details=true" % self.host
        response, info = fetch_url(self.module, url, method='GET')

        if info['status'] != 200:
            print(response)
            print(info)
            self.module.fail_json(
                msg="Request failed",
                status_code=info['status'],
                response=info['msg'],
                url=url)

        self.repo_list = json.load(response)
```

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
pulp_repo

##### ANSIBLE VERSION

```
ansible 2.8.2
```

##### STEPS TO REPRODUCE

```yaml
   pulp_repo:
      # connect / auth options
      pulp_host: REDACTED
      url_username: admin
      url_password: "{{ pulp_admin_password }}"
      force_basic_auth: true
      # actual repo options
      name: "SOME REPO NAME"
      relative_url: "/redhat/7/example"
      ca_cert: "/etc/rhsm/ca/redhat-uep.pem"
      client_cert: "/etc/pki/entitlement/999999999999999999.pem "
      client_key: "/etc/pki/entitlement/9999999999999999999-key.pem"
      feed: "https://cdn.redhat.com/content/dist/rhel/server/7/7Server/x86_64/os/"
      wait_for_completion: yes
```

##### EXPECTED RESULTS
create / modify of the pulp repo on the server

##### ACTUAL RESULTS
Connection failure: [SSL: SSL_HANDSHAKE_FAILURE] ssl handshake failure (_ssl.c:1822)

```paste below

```

### PR DESCRIPTION
**Pull request by coreaut**
##### SUMMARY
renamed the module parameters:
```
ca_cert
client_key
client_cert
```

because they wrongly overlapped with fetch_url allowed parameters.

Fixes #59513

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
pulp_repo

##### ADDITIONAL INFORMATION
To use a remote repo in pulp for sync which needs tls client auth (redhat repos) you need
to set this vars.
But they are also used in the ansible fetch_url function. So the module uses tls client auth also for API calls to the pulp server which is wrong.

It should only use tls client auth for the remote repo you wan't to sync. The Doku of the module says those vars are **only** for the remote repo.

### CODE REVIEW

**Comment by coreaut** [comment]
To better understand why the parameter names are wrong:

https://github.com/ansible/ansible/issues/59513#issuecomment-514631983

They have accidentally been renamed in #54315

**Review by sysadmind — APPROVED** [review]
Nice legwork on this one. I haven't used pulp in a while, but I think these changes are good barring any input from the project on renaming parameters.

**Comment by sysadmind** [comment]
shipit

**Comment by coreaut** [comment]
@sysadmind I think you need to answer in the referenced issue for someone to check and approve it?

**Comment by coreaut** [comment]
reopen because of stale_ci label.

**Comment by sysadmind** [comment]
I think you probably need to rebase this for the CI tests to pass. [A reference on rebase ](https://git-scm.com/book/en/v2/Git-Branching-Rebasing)

**Comment by sysadmind** [comment]
shipit

**Comment by coreaut** [comment]
bot_broken

thanks for the help, hope I'm not wrong but I think the bot is broken? It ignores the shipit command. Thanks for the help Ansible staff.

**Comment by jamescassell** [comment]
I'd drop version_added entirely, then

shipit

I think this should be backported since it's a bugfix and the module hasn't worked properly since 2.8.0 w/o this fix.

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (tremble) [inline lib/ansible/modules/packaging/os/pulp_repo.py:572]:
```suggestion
    importer_ssl_client_cert = module.params['feed_client_cert']
    if importer_ssl_client_cert is None and module.params['client_cert'] is not None:
      importer_ssl_client_cert = module.params['client_cert']
      module.deprecate("In Ansible 2.10 a new `feed_client_cert` option was added.  Until Ansible 2.14 the default value will come from the client_cert option", version="2.14")
```

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (tremble) [inline lib/ansible/modules/packaging/os/pulp_repo.py]:
```suggestion
        feed_ca_cert=dict(aliases=['importer_ssl_ca_cert','ca_cert']),
```

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (tremble) [inline lib/ansible/modules/packaging/os/pulp_repo.py]:
```suggestion
    aliases: [ importer_ssl_ca_cert, ca_cert ]
```

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (tremble) [inline lib/ansible/modules/packaging/os/pulp_repo.py:71]:
```suggestion
  feed_client_key:
    version_added: 2.10
```

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (tremble) [inline lib/ansible/modules/packaging/os/pulp_repo.py:58]:
```suggestion
  feed_client_cert:
    version_added: 2.10
```

**Review by tremble — CHANGES_REQUESTED** [review]
What I was describing on IRC...

You currently use ``argument_spec = url_argument_spec()``

This will automatically (re-) add client_cert and client_key to your argument_spec and the doc fragment will add them to the documentation.

My suggestion is as follows (changes beyond what you have today):

- add ca_cert as an alias to peer_ca_cert
- add the ``version_added: 2.10`` entries to the documentation
- For now, set the defaults based on the client_cert/client_key values

**Comment by coreaut** [comment]
> What I was describing on IRC...
> 
> You currently use `argument_spec = url_argument_spec()`
> 
> This will automatically (re-) add client_cert and client_key to your argument_spec and the doc fragment will add them to the documentation.
> 
> My suggestion is as follows (changes beyond what you have today):
> 
>     * add ca_cert as an alias to peer_ca_cert
> 
>     * add the `version_added: 2.10` entries to the documentation
> 
>     * For now, set the defaults based on the client_cert/client_key values

You mean feed_ca_cert not peer?
I did not know that client_cert and client_key get auto added. But that should be ok, one can configure pulp also to use tls client auth and not http basic auth.

I think however client_cert should not go into importer_ssl_client_cert if set.
Because it could override the certs one wants pulp to use to sync remote repos.

So a scenario like this would work as intended:
ansible -> tls client auth -> pulp server -> tls client auth -> remote repo

**Comment by tremble** [comment]
Yes, sorry I meant feed_ca_peer

> I think however client_cert should not go into importer_ssl_client_cert if set.

1. If you set feed_client_cert, this takes precedence.  Always.
2. If you don't set feed_client_cert, but do set client_cert, you get the same behaviour as 2.9
3. If you set feed_client_cert to "" then if won't default to client_cert and the various if clauses will ignore it

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (coreaut) [reply lib/ansible/modules/packaging/os/pulp_repo.py:572]: here's also a accidentally peer in the deprecate :)

**Comment by coreaut** [comment]
@tremble thanks for your time and help here. 
Can you change your suggestion to fix the typo about peer in the deprecation warning?
I'm not sure if you got a message about my comment on that one.

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (tremble) [inline lib/ansible/modules/packaging/os/pulp_repo.py:578]:
```suggestion
    importer_ssl_client_key = module.params['feed_client_key']
    if importer_ssl_client_key is None and module.params['client_key'] is not None:
      importer_ssl_client_key = module.params['client_key']
      module.deprecate("In Ansible 2.10 a new `feed_client_key` option was added.  Until Ansible 2.14 the default value will come from the client_key option", version="2.14")
```

**Review by coreaut — COMMENTED** [review]
keep ca_cert as alias and incorporate old behavior until it deprecates in 2.14, also include added_version info.

Fine by me and thanks!

**Comment by coreaut** [comment]
@tremble about the deprecation, before the 2.14 release I have to submit again to remove the then deprecated behavior?

**Comment by tremble** [comment]
As soon as 2.13 and 2.14 branch you would be able to do this.  I *believe* this should be in around 18 months time...

**Comment by coreaut** [comment]
@sysadmind @tremble @jamescassell @felixfontein 
I humbly request shipits after you have checked it, thanks.

It can only be backported after the fix in devel is out. 
[https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs](url)

**Comment by sysadmind** [comment]
shipit

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/pulp_repo.py]: This is not a good idea, since `ca_cert` is already a module option. (Due to `url_argument_spec()` above.)

Either do
```suggestion
        feed_ca_cert=dict(aliases=['importer_ssl_ca_cert'']),
```
and make sure that `feed_ca_cert` is set to `ca_cert`'s value if not set, or name the option `ca_cert` and add `feed_ca_cert` and `importer_ssl_ca_cert` as aliases. (Then there's no way to separate these options.)

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/pulp_repo.py:73]: You should mention that the default value is taken from `client_key` if that's specified, and that this will change in Ansible 2.14.

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/pulp_repo.py:60]: You should mention that the default value is taken from `client_cert` if that's specified, and that this will change in Ansible 2.14.

File: changelogs/fragments/59522-renamed-module-tls-client-auth-params-to-avoid-overlaping-with-fetch_url.yaml
Comment (felixfontein) [inline changelogs/fragments/59522-renamed-module-tls-client-auth-params-to-avoid-overlaping-with-fetch_url.yaml]: That doesn't help anyone understanding what happened without knowing some Ansible internals. How about something like:
```suggestion
  - pulp_repo - the ``client_cert`` and ``client_key`` options were used for both requests to pulp.io and for the repo to sync with, resulting in errors when they were used. Use the new options ``feed_client_cert`` and ``feed_client_key`` for client certificates that should only be used for repo synchronisation, and not for communication with pulp.io. (https://github.com/ansible/ansible/issues/59513)
```

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/pulp_repo.py]: How about:
```suggestion
        module.deprecate("To specify client certificates to be used with the repo to sync, and not for communication with pulp.io, use the new options `feed_client_cert` and `feed_client_key` (available since Ansible 2.10). Until Ansible 2.14, the default value for `feed_client_cert` will be taken from `client_cert` if only the latter is specified", version="2.14")
```
(And something similar below.)

**Comment by felixfontein** [comment]
If this should be backported, explicitly referencing Ansible 2.10 is not a good idea. That would be wrong, then. I'm not sure how to proceed here, though. Maybe it's best to merge it with 2.10 mentioned, change it to 2.9.x in the backport to `stable-2.9` (whatever the next 2.9.x version will be), and once that's merged, add a PR to `devel` which updates the version numbers accordingly.

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (tremble) [reply lib/ansible/modules/packaging/os/pulp_repo.py]: https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py#L1410

I may have missed it, I just don't see it...

Comment (coreaut) [reply lib/ansible/modules/packaging/os/pulp_repo.py]: @felixfontein I also don't see it being used by url_argument_spec()?

**Comment by coreaut** [comment]
> If this should be backported, explicitly referencing Ansible 2.10 is not a good idea. That would be wrong, then. I'm not sure how to proceed here, though. Maybe it's best to merge it with 2.10 mentioned, change it to 2.9.x in the backport to `stable-2.9` (whatever the next 2.9.x version will be), and once that's merged, add a PR to `devel` which updates the version numbers accordingly.

I will go with that and change it accordingly later if the backport gets accepted.

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (coreaut) [reply lib/ansible/modules/packaging/os/pulp_repo.py:60]: Thanks that's  a good idea. Added a commit for this.

**Comment by coreaut** [comment]
If no one has further suggestions I'm ready with this PR :)

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (felixfontein) [reply lib/ansible/modules/packaging/os/pulp_repo.py]: Hmm, I was pretty sure it was in there. Apparently, I'm wrong. Sorry!

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/pulp_repo.py]: I would add this to a new paragraph, so it will be easier to read (and harder to miss).
```suggestion
      - If not specified the default value will come from client_cert. Which will
```

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/pulp_repo.py]: Same here:
```suggestion
      - If not specified the default value will come from client_key. Which will
```

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/pulp_repo.py:545]: Cool, I didn't knew we had `deprecated_aliases`! Thanks @tremble!

**Review by felixfontein — COMMENTED** [review]
Some details, besides that LGTM!

File: lib/ansible/modules/packaging/os/pulp_repo.py
Comment (coreaut) [reply lib/ansible/modules/packaging/os/pulp_repo.py]: yes that's a good idea, thanks

**Review by felixfontein — APPROVED** [review]
shipit

**Comment by felixfontein** [comment]
@sysadmind @tremble still happy with this PR? If yes, I'll merge.

**Review by tremble — APPROVED** [review]

**Comment by tremble** [comment]
@felixfontein LGTM

**Comment by felixfontein** [comment]
If nobody complains I'll merge it on Monday, create a backport and put it on the project meeting agenda for Tuesday to see what the core team thinks about backporting this.

**Comment by felixfontein** [comment]
@coreaut thanks for fixing this!
@sysadmind @jamescassell @tremble thanks for reviewing this!

**Pull request merged**
